src/hotspot/share/memory/metaspace.cpp

```diff
@@ -627,11 +627,28 @@ ReservedSpace Metaspace::reserve_address_space_for_compressed_classes(size_t siz
     log_debug(metaspace, map)("Trying between " UINT64_FORMAT_X " and " UINT64_FORMAT_X
                               " with " SIZE_FORMAT_X " alignment", min, max, alignment);
     result = os::attempt_reserve_memory_between((char*)min, (char*)max, size, alignment, randomize);
+
+#ifdef AARCH64
+    if (result == nullptr) {
+      // If that failed, attempt to allocate at any properly aligned address. Try first an alignment that
+      // includes shift (e.g. 32GB). Failing that, try an alignment that does not include the shift (4GB).
```

Comment:
The way movk mode is currently implemented is to apply the right-shifted base to the narrow Klass ID, then to left-shift the thing to get Klass*

https://github.com/openjdk/jdk/blob/d0ea2a51111bd5de5a6465e7de6a4950aae89c71/src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp#L4723-L4731

which requires the *right-shifted* base to be movk-able. So it must have the lower 32+shift bits zero:

https://github.com/openjdk/jdk/blob/d0ea2a51111bd5de5a6465e7de6a4950aae89c71/src/hotspot/cpu/aarch64/macroAssembler_aarch64.cpp#L4656-L4659

hence the 32g alignment.

```
movw dst, src
movk dst, right-shifted-base to q3
shl dst, dst, 3
```

Note that this patch is still draft. I wanted to add another movk decoding variant, would work with a 4GB aligned base, just to increase our options. 

It would apply the unshifted base to the left-shifted narrow Klass ID:

```
shl dst, src, 3
movk dst, unshifted-base
```

and that would even need one less operation. 

Only problem is that it limits the value range of the narrow Klass ID, since it must not spill over into the third quadrant, so its upper (shift) bits must be zero. For shift=3, it means narrowKlass must be <= 2^(32-3). I think this is fine though, since we still can address 4 GB of class space. In other words, we should have never seen such high narrow Klass IDs anyway.

Reply:
> The way movk mode is currently implemented is to apply the right-shifted base to the narrow Klass ID, then to left-shift the thing to get Klass*

There's no need to left shift the Klass if we're using `movk` mode. Shift should be zero. We should always use a zero shift, 4G aligned, class space. That's how it was intended to work when I wrote it.